### DISCUSSION TITLE
Troubles with crawlWrap function

### DISCUSSION THREAD

**Discussion by AlineGiroux** [bmcclintock/momentuHMM#110] [Q&A] [answered]
I am having some troubles with crawlWrap function.
First of all, resampled location points do not appear to have realistic values of covariates associated with. 
Please see my example. 
I used two extremely regular tracks, and I compared the results with no resampling (case 1) and with resampling (case 2). 
Despite the results of "fit0" (with no covariates) are pretty similar in the two cases, the results of "fit 1" is very different to both covariates that I used. 

[table.txt](https://github.com/bmcclintock/momentuHMM/files/7498793/table.txt)

`#LOAD PACKAGE
library(momentuHMM)

#LOAD DATA
table<-read.table("table.txt",sep=",",h=T)
table$datetime <- as.POSIXct(paste(data$date,data$time),format="%d/%m/%Y %H:%M:%S")
head(table)

##################################################################
###############CASE 1: Using gps data with no resampling###################

#PREPARE MOVEMENT DATA
data.mov <- prepData(data=table,covNames=c("hour","tpt2"))
plot(data.mov)

#DEFINE MODES 
modes=c("encamping","moving")

#DEFINE INITIAL PRIORS TO FIT 0
#initial step distribution parameters (mean, standard deviation and zeromass)
step=c(1,40,4,100,0,0)
#initial angle distribution parameters (mean and concentration)
ang=c(1,0,2,0.1) 

#RUN FIT 0 MODEL USING THE PREVIOUSLY CHOSEN PRIORS
###(MODEL FIT 0 = SWITCH BEHAVIOR)###
fit0 = fitHMM(data=data.mov,nbStates=2,stateNames=modes,
              dist = list(step = "gamma", angle = "vm"),
              Par0 = list(step = step, angle = ang),
              estAngleMean = list(angle=TRUE))

plot(fit0)

#DEFINE PRIORS TO FIT 1
parameters1 <- getPar0(model=fit0, formula=~tpt2*cosinor(hour,24))

#RUN FIT 1 MODEL USING PRIORS BASED ON THE OUTPUT OF FIT 1 MODEL 
###(MODEL FIT 1 = SWITCH BEHAVIOR ~ HOUR OF DAY * TEMPERATURE)###
fit1 = fitHMM(data=data.mov, nbStates=2,stateNames=modes,
              dist = list(step = "gamma", angle = "vm"),
              Par0 = parameters1$Par,
              formula = ~tpt2*cosinor(hour,24),
              estAngleMean = list(angle=TRUE))

plot(fit1,plotCI=T,cov=data.frame(hour=0,tpt2=-2))

##################################################################
#############CASE 2: Using crawlWrap function to  resample gps data########### 

#FIT CONTINUOUS MOVEMENT MODEL 
cr<-crawlWrap(table,attempts=10,Time.name="datetime",coord=c("x","y"),timeStep="20 min",theta=c(10, .8),fixPar=c(NA,NA),ncores=3)
plot(cr)

#PREPARE MOVEMENT DATA
data.mov <- prepData(data=cr,covNames=c("hour","tpt2"))
plot(data.mov)

#DEFINE MODES 
modes=c("encamping","moving")

#DEFINE INITIAL PRIORS TO FIT 0
#initial step distribution parameters (mean and standard deviation)
step=c(1,40,4,100)
#initial angle distribution parameters (mean and concentration)
ang=c(1,0,2,0.1) 

#RUN FIT 0 MODEL USING THE PREVIOUSLY CHOSEN PRIORS
###(MODEL FIT 0 = SWITCH BEHAVIOR)###
fit0 = fitHMM(data=data.mov,nbStates=2,stateNames=modes,
              dist = list(step = "gamma", angle = "vm"),
              Par0 = list(step = step, angle = ang),
              estAngleMean = list(angle=TRUE))

plot(fit0)

#DEFINE PRIORS TO FIT 1
parameters1 <- getPar0(model=fit0, formula=~tpt2*cosinor(hour,24))

#RUN FIT 1 MODEL USING PRIORS BASED ON THE OUTPUT OF FIT 1 MODEL 
###(MODEL FIT 1 = SWITCH BEHAVIOR ~ HOUR OF DAY * TEMPERATURE)##
fit1 = fitHMM(data=data.mov, nbStates=2,stateNames=modes,
              dist = list(step = "gamma", angle = "vm"),
              Par0 = parameters1$Par,
              formula = ~tpt2*cosinor(hour,24),
              estAngleMean = list(angle=TRUE))

plot(fit1,plotCI=T,cov=data.frame(hour=0,tpt2=-2))

`

Please let me know if I am doing something wrong. Thanks.

**Comment by bmcclintock** [comment, marked as answer]
You have time-dependent covariates corresponding to the times of the observations, but when you "resample" the tracks at 20min intervals, neither `crawlWrap` nor `prepData` know what the covariate values should be for the predicted locations. If you inspect `cr$crwPredict` you'll see all of the missing covariate values at the prediction times:
```
head(cr$crwPredict[c("ID","datetime","locType","mu.x","mu.y","hour","tpt2")])
    ID            datetime locType     mu.x    mu.y hour     tpt2
1 kiko 2016-04-17 12:21:29       p 626431.0 7870892   12 2.205735
3 kiko 2016-04-17 12:41:29       p 626528.9 7870817   NA       NA
4 kiko 2016-04-17 12:41:32       o 626529.0 7870817   12 2.205735
5 kiko 2016-04-17 13:01:29       p 626498.4 7870847   NA       NA
6 kiko 2016-04-17 13:21:27       o 626437.0 7870896   13 2.252425
7 kiko 2016-04-17 13:21:29       p 626436.9 7870896   NA       NA
```
 `prepData` attempts to fill these in using the closest non-missing values, and you should have received the following warning from `prepData`:
```
Warning message:
In prepData.default(data = cr, covNames = c("hour", "tpt2")) :
  There are 27588 missing covariate values. Each will be replaced by the closest available value.
```
You can see the resulting values for the covariates are complete nonsense:
```
plot(data.mov,dataNames=c("hour","tpt2"))
```
![image](https://user-images.githubusercontent.com/11667599/140793453-f2dcfa20-aedc-4ce7-aa79-f943b0597c61.png)
![image](https://user-images.githubusercontent.com/11667599/140793479-e3451468-9577-4ca5-96a8-b04e4181772d.png)
So the bottom line is that when predicting locations at times other than those observed, any non-spatial time-dependent covariate values need to be manually specified (e.g., see `?crawlMerge`) before using `prepData` and `fitHMM`.

**Comment by AlineGiroux** [comment]
Indeed. Thank you very much.
Problem solved.
